src/main/java/com/thealgorithms/datastructures/heaps/FibonacciHeap.java

```diff
@@ -0,0 +1,436 @@
+    				tmpMin.setNext(buckets[i]);
+    				if (buckets[i].getKey() < tmpMin.getKey()) {
+    					tmpMin= buckets[i];
+    				}
+    			}
+    		}
+    	}
+    	return tmpMin;
+    }
+
+    /*
+     * link between two nodes (and their trees)
+     * defines the smaller node to be the parent
+     */
+    private HeapNode link(HeapNode c1, HeapNode c2) {
+    	if (c1.getKey()> c2.getKey()) {
+    		HeapNode c3 = c1;
+    		c1 = c2;
+    		c2 = c3;
+    	}
+    	if (c1.child == null) {
+    		c1.child = c2;
+    	}
+    	else {
+    		c1.child.setNext(c2);
+    	}
+    	c2.parent = c1;
+		c1.rank++;
+		totalLinks++;
+    	return c1;
+    }
+
+
+   /**
+    * public class HeapNode
+    * each HeapNode belongs to a heap (Inner class)
+    */
+    public class HeapNode{
+
+    	public int key;
+    	private int rank;
+    	private boolean marked;
+    	private HeapNode child;
+    	private HeapNode next;
+    	private HeapNode prev;
+    	private HeapNode parent;
+
+    	/*
+    	 * a constructor for a heapNode withe key @param (key)
+    	 * prev == next == this
+    	 * parent == child == null
+    	 */
+    	public HeapNode(int key) {
+    		this.key = key;
+    		this.marked = false;
+    		this.next = this;
+    		this.prev = this;
+    	}
+
+    	/*
+    	 * returns the key of the node.
+    	 */
+    	public int getKey() {
+    		return this.key;
+    	}
+
+
+    	/*
+    	 * checks whether the node is marked
+    	 * $ret = true if one child has been cut
+    	 */
+    	private boolean isMarked() {
+    		return this.marked;
+    	}
+
+    	/*
+    	 * mark a node (after a child was cut)
+    	 * @inv root.mark() == false.
+    	 */
+    	private void mark() {
+    		if (this.isRoot()) {return;}  //check if the node is a root
+    		this.marked = true;
+    	}
+
+    	/*
+    	 * add the node @param (newNext) to be between this and this.next
+    	 * works fine also if @param (newNext) does not "stands" alone
+    	 */
+    	private void setNext(HeapNode newNext) {
+    		HeapNode tmpNext = this.next;
+    		this.next = newNext;
+    		this.next.prev.next = tmpNext;
+    		tmpNext.prev = newNext.prev;
+    		this.next.prev = this;
+    	}
+
+    	/*
+    	 * returns the next node to this node
+    	 */
+    	private HeapNode getNext() {
+    		return this.next;
+    	}
+
+    	/*
+    	 * check if the node is a root
+    	 * root definition - this.parent == null (uppest in his tree)
+    	 */
+    	private boolean isRoot() {
+    		return (this.parent == null);
+    	}
+    }
+
+    public static void main(String[] args) {
+        FibonacciHeap fibonacciHeap = new FibonacciHeap();
+        fibonacciHeap.insert(5);
+        HeapNode node = fibonacciHeap.insert(3);
+        fibonacciHeap.insert(1);
+        fibonacciHeap.insert(18);
+        fibonacciHeap.insert(33);
+
+        System.out.println(fibonacciHeap.findMin().getKey());
+        fibonacciHeap.deleteMin();
+        System.out.println(fibonacciHeap.findMin().getKey());
+
+    }
```

Comment:
Please replace this with tests using `@Test`, see examples in the codebase in the `src/test` directory